pkg/queue/breaker_test.go

```diff
@@ -133,28 +110,74 @@ func TestBreakerLargeCapacityRecover(t *testing.T) {
 	}
 	// Check the breaker reset
 	if !reflect.DeepEqual(want[len(want)-10:], got[len(got)-10:]) {
-		t.Fatalf("Wanted %v. Got %v.", want, got)
+		t.Fatalf("Wanted\n%v.\nGot\n%v.", want, got)
 	}
 }
 
-func (b *Breaker) concurrentRequest() (chan struct{}, chan bool) {
-	release := make(chan struct{})
-	thunk := func() {
-		_, _ = <-release
+func TestUnlimitedBreaker(t *testing.T) {
+	b := NewBreaker(1, 0)
+	requests := b.concurrentRequests(1000)
+	unlockAll(requests)
+	for i, ok := range accepted(requests) {
+		if !ok {
+			t.Fatalf("Expected request %d to be successful, but it failed.", i)
+		}
 	}
-	result := make(chan bool)
+}
+
+// Attempts to perform a concurrent request against the specified breaker.
+func (b *Breaker) concurrentRequest() request {
+
+	// There is a brief window between when capacity is released and
+	// when it becomes available to the next request.  We yield here
+	// to reduce the likelihood that we hit that edge case.  E.g.
+	// without yielding `go test ./pkg/queue/breaker.* -count 10000`
+	// will fail about 3 runs.
+	runtime.Gosched()
+
+	r := request{lock: &sync.Mutex{}, accepted: make(chan bool, 2)}
+	r.lock.Lock()
+	started := make(chan bool)
 	go func() {
-		result <- b.Maybe(thunk)
+		started <- true
+		ok := b.Maybe(func() {
+			r.lock.Lock() // Will block on locked mutex.
+			r.lock.Unlock()
+		})
+		r.accepted <- ok
 	}()
-	runtime.Gosched()
-	return release, result
+	<-started // Ensure that the go func has had a chance to execute.
```

Comment:
I'm not really convinced this channel does anything...

It seems that the absence of `close(started)` is fine: https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open

Reply:
also `sync.WaitGroup` may be clearer than using channels.

Reply:
I've gotten that feedback from another developer too.  The channel implementation seems the most intuitive to me, but I think the imperative wait group approach is more generally readable.

Changed to a wait group.